static/js/lib/auth.js

```diff
@@ -24,9 +28,78 @@ export const STATE_REGISTER_CONFIRM = "register/confirm"
 export const STATE_REGISTER_DETAILS = "register/details"
 export const STATE_REGISTER_EXTRA_DETAILS = "register/extra"
 
+export const ALL_STATES = [
+  STATE_ERROR,
+  STATE_ERROR_TEMPORARY,
+  STATE_SUCCESS,
+  STATE_INACTIVE,
+  STATE_INVALID_EMAIL,
+  STATE_USER_BLOCKED,
+  STATE_LOGIN_EMAIL,
+  STATE_LOGIN_PASSWORD,
+  STATE_LOGIN_PROVIDER,
+  STATE_REGISTER_EMAIL,
+  STATE_REGISTER_CONFIRM,
+  STATE_REGISTER_CONFIRM_SENT,
+  STATE_REGISTER_DETAILS,
+  STATE_REGISTER_EXTRA_DETAILS
+]
+
 export const generateLoginRedirectUrl = () => {
   const { pathname, search, hash } = window.location
 
   const next = `${pathname}${search}${hash}`
   return `${routes.login.begin}?${qs.stringify({ next })}`
 }
+
+export type StateHandlers = {
+  [AuthStates]: (response: AuthResponse) => void
+}
+
+const getErrorQs = (errors: Array<string>) =>
+  !isEmpty(errors)
+    ? qs.stringify({
+      error: errors[0]
+    })
+    : ""
+
+export const handleAuthResponse = (
+  history: RouterHistory,
+  response: AuthResponse,
+  handlers: StateHandlers
+) => {
+  /* eslint-disable camelcase */
+  const { state, redirect_url, partial_token, errors, field_errors } = response
+
+  // if this has a handler keyed off the state name invoke it
```

Comment:
Possibly clearer: `If a specific handler function was passed in for this response state, invoke it`